sdk/protocol-plugins/src/spark/SparkProtocolPlugin.ts

```diff
@@ -1,18 +1,234 @@
-import { ActionBuilder, ActionBuildersMap } from '@summerfi/order-planner-common/builders'
-import { SimulationSteps, steps } from '@summerfi/sdk-common/simulation'
-import { SparkDepositBorrowActionBuilder } from './builders'
-import { Maybe } from '@summerfi/sdk-common/common'
-import { ProtocolName } from '@summerfi/sdk-common/protocols'
-import { IProtocolActionBuilder } from '@summerfi/order-planner-common/interfaces'
-
-export class SparkProtocolPlugin implements IProtocolActionBuilder {
-  readonly protocol = ProtocolName.Spark
-
-  readonly StepBuilders: Partial<ActionBuildersMap> = {
-    [SimulationSteps.DepositBorrow]: SparkDepositBorrowActionBuilder,
+import {
+  AddressValue,
+  Percentage,
+  TokenAmount,
+  TokenSymbol,
+  Price,
+  CurrencySymbol,
+  RiskRatio,
+  Position,
+  ChainId,
+  ChainFamilyName,
+  valuesOfChainFamilyMap,
+} from '@summerfi/sdk-common/common'
+import { SimulationSteps } from '@summerfi/sdk-common/simulation'
+import type { SparkPoolId } from '@summerfi/sdk-common/protocols'
+import { PoolType, ProtocolName, EmodeType } from '@summerfi/sdk-common/protocols'
+import { BigNumber } from 'bignumber.js'
+import { z } from 'zod'
+import { SparkDepositBorrowActionBuilder } from './builders/SparkDepositBorrowActionBuilder'
+import { BaseProtocolPlugin } from '../implementation/BaseProtocolPlugin'
+import { IPositionId } from '../interfaces/IPositionId'
+import { SparkPoolCollateralConfig, SparkLendingPool, SparkPoolDebtConfig } from './Types'
+import {
+  AaveV3LikePluginBuilder,
+  filterAssetsListByEMode,
+} from '../implementation/AAVEv3LikeBuilder'
+import { UNCAPPED_SUPPLY, PRECISION_BI } from '../implementation/constants'
+
+export class SparkProtocolPlugin extends BaseProtocolPlugin<SparkPoolId> {
+  public static protocol: ProtocolName.Spark = ProtocolName.Spark
+  public static supportedChains = valuesOfChainFamilyMap([ChainFamilyName.Ethereum])
+  public static schema = z.object({
+    protocol: z.object({
+      name: z.literal(ProtocolName.Spark),
+      chainInfo: z.object({
+        name: z.string(),
+        chainId: z.custom<ChainId>(
+          (chainId) =>
+            SparkProtocolPlugin.supportedChains.some((chainInfo) => chainInfo.chainId === chainId),
+          'Chain ID not supported',
+          true,
+        ),
+      }),
+    }),
+    emodeType: z.nativeEnum(EmodeType),
+  })
+
+  constructor() {
+    const StepBuildersMap = {
+      [SimulationSteps.DepositBorrow]: SparkDepositBorrowActionBuilder,
+    }
+    super(
+      SparkProtocolPlugin.protocol,
+      SparkProtocolPlugin.supportedChains,
+      SparkProtocolPlugin.schema,
+      StepBuildersMap,
+    )
   }
 
-  getActionBuilder<T extends steps.Steps>(step: T): Maybe<ActionBuilder<T>> {
-    return this.StepBuilders[step.type] as ActionBuilder<T>
+  async getPool(poolId: unknown): Promise<SparkLendingPool> {
```

Comment:
This method is too long, we should divide it in sub-functions with sensible names to make it more readable. The loop on filtered asset list could call a function like `getCollateralAssetInfo` that returns the price, priceUSD, etc... That way you can also change the `for` loop for a `filteredAssetsList.map()` which flows better and eliminates the need to declare the `collaterals` record empty

Reply:
Yea, no issue. I agree.